### ISSUE TITLE
JsPsych Record Video

### ISSUE THREAD

**Issue by okaycj** [lookit/lookit-jspsych#35]
# TL;DR
Record video of participant and save it to AWS S3 store. 

# Narrative
EFP currently has the option of recording participants while they work through a study.  Additionally, it will record a consent video.  Both of these videos are saved in a S3 store and a database entry is added to the Video table.  These two features and a trial type for setting up their camera and microphone will need to be added to the CHS jsPsych NPM package that was created from lookit/lookit-api#1282.

# Acceptance Criteria
- [x] Documentation: #32
- [x] Tests: #25
- [x] Record video/audio: #25
- [x] Save file to S3 Store via multipart upload: #29
- [x] Database entry for video file: #36
- [x] Custom plugin/extension for consent trial, start/stop video recording, video config trial: #25
- [ ] Recorder timestamps that give video recording timestamps (time since the start of recording) that correspond to trial events (start/load, others?)
- [ ] Default functionality and parameters for determining what happens while (a) waiting for the recording to initialize and (b) waiting for the recording to finish uploading. This includes time limits and content to display to the participant.
- [ ] Error catching/logging for upload timeouts and other uploading problems
- [ ] Try to stop (abort or complete) in-progress recordings if participant exits study early
- [x] Ability to playback recordings, e.g. in EFP video-consent frame - #52
- [x] Ability to re-record a video (and save additional video rather than overwrite), e.g. in EFP video-consent frame - #40, #52
- [x] Check for cam/mic access, mic volume, and media permissions, e.g. in video-config frame - #45

# Implementation Notes
We'd like to remove the difference between trial video recording and session recording.  It would be nice to have it all be the same thing.  The user should be able to start/stop the video at their whim.

jsPsych has [initialize camera](https://www.jspsych.org/7.3/plugins/initialize-camera/) and [initialize microphone](https://www.jspsych.org/7.3/plugins/initialize-microphone/) plugins, as well as a [record-video extension](https://www.jspsych.org/7.3/extensions/record-video/), which allows users to use video recording during any trial (plugin). We can use the initialize camera/mic plugin code as part of a custom camera/mic set-up plugin (similar to EFP `video-config`). We may not need to do much more customization for the video recording extension, other than setting file names and saving the data.

## Edit: Notes (below) from duplicate issue

We are implementing this in a lookit-jspsych `record` package that contains the following:
- Trial-level recording for any jsPsych trials with a CHS Record extension
- Session recording for any set of jsPsych trials with CHS Record `start` and `stop` plugins, which start/stop the session recording

**Comment by becky-gilbert** [comment]
Closing this as completed, since we have working video recording in jsPsych. Moving the remaining to-do items into new issues.
